mongomock/collection.py

```diff
@@ -826,6 +795,77 @@ def _copy_only_fields(self, doc, fields, container):
                 fields[field] = op
             return doc_copy
 
+    def _subdocitem_merge(self, doc_copy, field, sdocitems, subtract=False):
+        sdoc_copy = doc_copy.setdefault(field, [])
+        if len(sdoc_copy) < len(sdocitems):
+            sdoc_copy.extend([{} if not subtract else {'$EMPTY$': True}
+                              for _ in xrange(len(sdocitems) - len(sdoc_copy))])
+        for (sdoc, sdocitem) in zip(sdoc_copy, sdocitems):
+            if not subtract:
+                sdoc.update(sdocitem)
+            else:
+                if sdoc == {'$EMPTY$': True}:
+                    sdoc.clear()
+                    sdoc.update(sdocitem)
+                else:
+                    exc_keys = set(sdoc.keys()).difference(sdocitem.keys())
+                    for key in exc_keys:
+                        del sdoc[key]
+
+    def _filter_subdocuments(self, doc, fields, container):
+        # if 1 was passed in as the field values, include those fields
```

Comment:
Comments above an `if` statement, where each suite holds a big chunk of code usually indicates a code-smell, where each suite should be a separate function or method. In this case additionally both suites look very similar (a.k.a copy-pasted), so that I think they should be refactored to a function. I know that you just adapted the code that was there before, but it would be nice if you could remove the obvious flaws while adding functionality.